### PR TITLE
Week 6 - Generate Flight Plan

### PR DESCRIPTION
**Pull request by devenshah2018**
**EDIT (10/20/25):**

I implemented a CEM optimizer for the generated flight plan to suggest configurations that would give either a better image quality or mission speed (user-specified). My notebook documents an example that performs the CEM optimizer on the nominal dataset, weighing image quality as 70% and mission speed as 30%. Note that this optimizer is still _experimental_.

---

The following changes implement core flight planning logic, add advanced modeling for camera angles and motion constraints, and improve documentation:

**Core Flight Planning Algorithms:**
- Implemented the `generate_photo_plan_on_grid` function to generate waypoints in a lawn-mower pattern, ensuring area coverage with specified overlaps and blur-free photo speed constraints.
- Added `compute_segment_travel_time` and `compute_plan_time` functions for realistic mission time estimation using trapezoidal and triangular speed profiles, accounting for acceleration and speed limits.

**Camera Modeling and Motion Blur Prevention:**
- Updated `compute_ground_sampling_distance` to support non-nadir (tilted) camera angles, ensuring correct GSD calculation and motion blur prevention for advanced scanning patterns.
- Enhanced `compute_speed_during_photo_capture` to accept camera angle as a parameter for accurate speed computation under non-nadir conditions.

**Data Model Improvements:**
- Expanded the `Waypoint` class to include full initialization, string representation, and optional look-at points for non-nadir scans, supporting advanced flight patterns and debugging.

**Documentation:**
- Added a detailed `README.md` and refactored notebook verbiage, covering system features, technical details, project structure, and acknowledgements, providing a clear overview and usage guide.

### CODE REVIEW

File: src/camera_utils.py [hunk 1]
```diff
@@ -108,24 +108,37 @@ def compute_image_footprint_on_surface(
...
     
-    # Get the image footprint
-    footprint = compute_image_footprint_on_surface(camera, distance_from_surface)
+    # Get the image footprint (accounting for camera angle if non-zero)
+    if camera_angle_rad == 0.0:
+        footprint = compute_image_footprint_on_surface(camera, distance_from_surface)
+    else:
+        # Import here to avoid circular dependency
+        from src.plan_computation import compute_non_nadir_footprint
+        footprint = compute_non_nadir_footprint(camera, distance_from_surface, camera_angle_rad)
```
Comment (ayushbaid) [inline src/camera_utils.py]: I guess compute_non_nadir_footprint for angle input = 0 does not return the same value as compute_image_footprint_on_surface because of the min operation we are doing?

File: src/camera_utils.py [hunk 2]
```diff
@@ -108,24 +108,37 @@ def compute_image_footprint_on_surface(
...
     
     # Calculate GSD in both directions
     gsd_x = footprint[0] / camera.image_size_x
     gsd_y = footprint[1] / camera.image_size_y
     
-    # Return the smaller GSD (higher resolution)
+    # For non-nadir angles, use the GSD in the direction of tilt (Y direction)
```
Comment (ayushbaid) [inline src/camera_utils.py:155]: Nice work documenting your decision.

File: src/data_model.py [hunk 3]
```diff
@@ -61,6 +61,34 @@ def __str__(self) -> str:
 class Waypoint:
     """
     Waypoints are positions where the drone should fly to and capture a photo.
+    
+    Attributes:
+        x (float): X coordinate in meters (eastward direction)
+        y (float): Y coordinate in meters (northward direction)
+        z (float): Z coordinate in meters (altitude above ground)
+        speed (float): Maximum speed at this waypoint in m/s for blur-free photos
+        look_at_x (float, optional): X coordinate of look-at point for non-nadir scans
```
Comment (ayushbaid) [inline src/data_model.py]: You went with the look-at point. Interesting!

File: src/plan_computation.py [hunk 4]
```diff
@@ -122,4 +126,166 @@ def generate_photo_plan_on_grid(
...
+        actual_dx = dataset_spec.scan_dimension_x / (num_images_x - 1)
+    else:
+        actual_dx = 0
+        
+    if num_images_y > 1:
+        actual_dy = dataset_spec.scan_dimension_y / (num_images_y - 1)
+    else:
+        actual_dy = 0
+    
+    # Calculate starting positions to center the grid
```
Comment (ayushbaid) [inline src/plan_computation.py]: Oh you assumed the origin was at the center of the scan dimension

File: src/plan_computation.py [hunk 5]
```diff
@@ -122,4 +126,166 @@ def generate_photo_plan_on_grid(
...
+    num_images_y = math.ceil(dataset_spec.scan_dimension_y / dy) + 1
+    
+    # Calculate the actual spacing to center the grid within the scan area. If we need more images than minimum, spread them evenly.
+    if num_images_x > 1:
+        actual_dx = dataset_spec.scan_dimension_x / (num_images_x - 1)
+    else:
+        actual_dx = 0
+        
+    if num_images_y > 1:
+        actual_dy = dataset_spec.scan_dimension_y / (num_images_y - 1)
```
Comment (ayushbaid) [inline src/plan_computation.py:131]: I think you are running into integer division here. That is why your waypoints are integers too. I was expecting distance between waypoints to be ~15.16m.

File: src/plan_computation.py [hunk 6]
```diff
@@ -122,4 +126,166 @@ def generate_photo_plan_on_grid(
...
+) -> Tuple[float, Dict[str, Any]]:
+    """
+    Compute minimum time to travel `distance` starting & ending at v_photo,
+    given max speed v_max and max accel a_max.
+
+    Returns (time_s, profile_dict).
+    profile_dict includes 'type' ('triangular'|'trapezoidal') and relevant params.
+    """
+    if distance <= 0.0:
+        return 0.0, {"type": "degenerate", "distance": distance}
```
Comment (ayushbaid) [inline src/plan_computation.py]: Nice handling here. Nit: would recommend using enum for type.

File: src/plan_computation.py [hunk 7]
```diff
@@ -122,4 +126,166 @@ def generate_photo_plan_on_grid(
...
+        # triangular profile (no cruise)
+        t_acc = (v_peak - v_photo) / a_max
+        total_time = 2.0 * t_acc
+        return total_time, {"type": "triangular", "v_peak": v_peak, "t_acc": t_acc}
+
+    # trapezoidal: reach v_max, cruise, then decel
+    v_cruise = v_max
+    # distance used accelerating from v_photo to v_cruise and back
+    d_accdec = (v_cruise * v_cruise - v_photo * v_photo) / a_max
+    if d_accdec >= distance:
```
Comment (ayushbaid) [inline src/plan_computation.py:193]: I agree this should not happen mathematically. If we reach there, it would mean a bug, or a numerical issue. Maybe we can assert here, but its fine as-is too.

Did you actually see any example where we end up here?

File: src/plan_computation.py [hunk 8]
```diff
@@ -122,4 +126,166 @@ def generate_photo_plan_on_grid(
...
+            pts.append(np.asarray(p))
+
+    total_time = 0.0
+    segments = []
+    n = len(pts)
+    if n <= 1:
+        return total_time, segments
+
+    # include initial photo dwell
+    total_time += exposure_time_s
```
Comment (ayushbaid) [inline src/plan_computation.py:241]: This is excellent work including the exposure time here too. A small nit is that the drone will still move during exposure, which we should cut from the next segment. But its small enough to ignore and not add more complexity here.

File: src/plan_optimizer.py
Comment (ayushbaid) [inline src/plan_optimizer.py:1]: This is excellent work! I did not review it too closely as you know more about this than me, but it looks good!

File: main.ipynb
Comment (ayushbaid) [inline main.ipynb:1]: The plotly figures are not rendered on github. I recommend that you save the plot and add it as an image to the notebook.

**Review by ayushbaid — APPROVED** [review]
Excellent work! Your idea of CEM is pretty cool, and you went over and above to implement that. 

However, I see that the optimal parameters computed has the heigh of 5m. That is super low and would be infeasible for most common scenarions. I recommend you cap the min and max gsd, and same for speed. 

I have few minor comments:
1. Add static images for your plots so that people browsing your repo can review it
2. Fix use of `_` in Latex. E.g. `image\_size`

File: src/camera_utils.py (hunk 1 above)
Comment (devenshah2018) [reply src/camera_utils.py]: It looks like the value of compute_non_nadir_footprint (where angle is set to 0) is equal to the value of compute_image_footprint_on_surface (nadir):

image

I tested with removing this if-condition, only using `footprint = compute_non_nadir_footprint(camera, distance_from_surface, camera_angle_rad)` and it had no effect on the notebook results. Will fix 👍

File: src/data_model.py (hunk 3 above)
Comment (devenshah2018) [reply src/data_model.py]: This was actually for a previous attempt, it is no longer used in this implementation. I will go ahead and remove it for simplicity.

File: src/plan_computation.py (hunk 5 above)
Comment (devenshah2018) [reply src/plan_computation.py:131]: I did some further digging and I found my `actual_dy` to be 10.71:

image

`actual_dx` uses the same formula as `actual_dy`, which is `scan_dim_x` / (`num_images_x` - 1) = 150 / (11 - 1) = 15.00 (without integer rounding). 

Using `dx` directly and not using `actual_dx` at all gives the expected 15.16, should I remove the `actual_dx` logic?

**Note**: This output is using the adjustment where I previously assumed the origin being the center of the scan dimension. I have updated my code locally where:

```python
if num_images_x > 1:
        actual_dx = dataset_spec.scan_dimension_x / (num_images_x - 1)
    else:
        actual_dx = start_x # instead of 0
        
    if num_images_y > 1:
        actual_dy = dataset_spec.scan_dimension_y / (num_images_y - 1)
    else:
        actual_dy = start_y # instead of 0
```

File: src/plan_computation.py (hunk 6 above)
Comment (devenshah2018) [reply src/plan_computation.py]: Thank you! Will add the enum 👍

File: src/plan_computation.py (hunk 7 above)
Comment (devenshah2018) [reply src/plan_computation.py:193]: **TLDR**: From a real-life standpoint it is highly unlikely we reach here, but not impossible. The check handles the edge case where the math says we can reach `v_max`, but the segment is too short to actually do so.

The reason I added this one is because `v_peak > v_max` checks in theory I can reach `v_max` if the segment is long enough. `d_accdec` checks the actual distance required to accelerate from `v_photo` to `v_max`. If my segment is shorter than `d_accdec` by directly checking, it is grounded in truth that I can't reach `v_max` before needing to decelerate (even if `v_peak > v_max`). 

Example:
- `v_photo` = 3
- `distance` = 50
- `v_max` = 16
- `a_max` = 3.5

`v_peak` = 13.56 -> `v_peak < v_max` -> triangle profile is used

If `distance` = 80 -> `v_peak` = 17 -> `v_peak > v_max` -> possible to reach `v_max` -> trapezoid

`d_accdec` = 70.6 -> if distance = 70 -> `d_accdec > distance` -> cannot reach `v_max` (even though `v_peak > v_max`) -> should be triangle but will return trapezoid

File: src/plan_computation.py (hunk 8 above)
Comment (devenshah2018) [reply src/plan_computation.py:241]: Awesome, thats good to know! I will hold off implementing it in the notebook, but possibly may show it in my interface.

File: src/plan_optimizer.py
Comment (devenshah2018) [reply src/plan_optimizer.py:1]: I appreciate that! There is still a lot of room for improvement, right now it is a very base case (quality metric is determined simply by the gsd). I am curious to play around with some of these metrics and different weights to see how much it can be optimized using CEM. 

This is very new to me as well, I originally tried using neural networks but backprop seemed like an overdevelopment if its possible to use CEM. I will add some markdown making it more clear that this is purely experimental.

File: src/plan_computation.py (hunk 4 above)
Comment (devenshah2018) [reply src/plan_computation.py]: I updated my code locally where:
```python
if num_images_x > 1:
        actual_dx = dataset_spec.scan_dimension_x / (num_images_x - 1)
    else:
        actual_dx = start_x # instead of 0
        
    if num_images_y > 1:
        actual_dy = dataset_spec.scan_dimension_y / (num_images_y - 1)
    else:
        actual_dy = start_y # instead of 0
```
